apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,401 @@
+  <img src="/posts/python-wheels-from-tags-to-variants/jax.png"
+    width="790" height="342" alt="A table listing supported JAX versions.
+The rows are CPU, NVIDIA GPU, Google Cloud TPU, AMD GPU, Apple GPU
+and Intel GPU. The columns are x86_64 and aarch64 Linux, aarch64 Mac
+and x86_64 Windows and Windows WSL2. The table cells indicate either 'yes'
+or 'experimental' with links, 'no' or 'n/a'." />
+  <figcaption>Fig. JAX version chooser</figcaption>
+</figure>
+</div>
+
+Let's look at another example. [JAX's "Supported Platforms" matrix](https://docs.jax.dev/en/latest/installation.html#supported-platforms) provides links to instructions for different installation options.
+Most of them suggest installing the [jax package](https://pypi.org/project/jax/)
+with an extra, e.g.: `jax[cuda12]`. And indeed, JAX publishes support
+for different variants as plugins that can be pulled in automatically
+via dependencies. This avoids using additional indexes, but still requires
+manual action from the user.
+
+Other variations of the same solution are possible. Rather than using plugins,
+one could publish different package variants using separate package names
+and use a metapackage with extras to select between them. Or simply ask the user
+to install the correct package. However, this has the unfortunate side effect
+that once multiple different variants end up being installed, they may overwrite
+one another.
+
+Or one can just publish a single package with all possible variants inside —
+except it may end up being huge.
+
+## Plugins to the rescue!
+
+As I have already pointed out, the solutions putting variant selection entirely
+within the package manager weren't very flexible. They could really work only
+if either the list of valid variants was largely fixed or at least predictable.
+To sidestep this limitation, we need a more distributed architecture, one where
+each package can independently devise how to select between its variants.
+This is how we reached an architecture based on plugins.
+
+<div style={{ textAlign: "center" }}>
+<figure style={{width: 'auto', margin: '0 2em', display: 'inline-block', verticalAlign: 'top'}}>
+  <img src="/posts/python-wheels-from-tags-to-variants/variant-flow.png"
+    width="683" height="577" alt="A flow chart showing interaction between
+wheel-related software and provider plugins. A PEP 517 backend (such
+as flit_core, hatchling, meson-python…) produces a Wheel Variant, which is
+then consumed by a wheel installer (pip, uv, pdm…). Alongside them, variant
+providers (CUDA, ROCm, x86_64, AArch64, BLAS / LAPACK, MPI, OpenMP…) are shown.
+PEP 517 backends communicate with them to validate variants, wheras installers
+do to get supported properties." />
+  <figcaption>Fig. Interaction between wheel software and provider plugins</figcaption>
+</figure>
+</div>
+
+While PEP 517 backends remain responsible for building wheels, and installers
```

Comment:
Minor editorial comment: I'd use "build backends" as the proper name, and put (PEP 517) in brackets behind it the first time the term build backends is used.  Both because that's the more common name and because there's a push in Python packaging land to refer to things not by a PEP number.

Reply:
Interesting, I wasn't aware of this. In fact, I've always been using "PEP 517" as a more specific kind of backend. I'll update it.